// isom-fetch/lib/defaults.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getUrlState = exports.setUrlState = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DEFAULT_CONTENT_TYPE = {
  'Content-Type': 'application/x-www-form-urlencoded'
};

var urlState = '__FETCH_URLS__';

exports.default = {
  headers: {
    common: {
      Accept: 'application/json, text/plain, */*'
    },
    patch: (0, _extends3.default)({}, DEFAULT_CONTENT_TYPE),
    post: (0, _extends3.default)({}, DEFAULT_CONTENT_TYPE),
    put: (0, _extends3.default)({}, DEFAULT_CONTENT_TYPE)
  },

  timeout: 0,

  xsrfCookieName: 'XSRF-TOKEN',
  xsrfHeaderName: 'X-XSRF-TOKEN',

  maxContentLength: -1,

  validateStatus: function validateStatus(status) {
    return status >= 200 && status < 300;
  }
};
var setUrlState = exports.setUrlState = function setUrlState(name) {
  (0, _invariant2.default)(typeof name === 'string', 'urlState 只能为字符串');

  urlState = name;
};

var getUrlState = exports.getUrlState = function getUrlState() {
  return urlState;
};
