// isom-fetch/lib/mask.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResponse = exports.getRequest = exports.getURL = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _defaults = require('./defaults');

var _defaults2 = _interopRequireDefault(_defaults);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function combineURLs(baseURL, relativeURL) {
  return baseURL.replace(/\/+$/, '') + '/' + relativeURL.replace(/^\/+/, '');
}

function isAbsoluteURL(url) {
  return (/^([a-z][a-z\d\+\-\.]*:)?\/\//i.test(url)
  );
}

var getURL = exports.getURL = function getURL(config) {
  if (config.baseURL && !isAbsoluteURL(config.url)) {
    return combineURLs(config.baseURL, config.url);
  }
  return config.url || '';
};

var getRequest = exports.getRequest = function getRequest() {
  var defConfig = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
  var config = arguments[1];

  config = (0, _extends3.default)({}, _defaults2.default, defConfig, config);

  config.url = getURL(config);

  config.withCredentials = config.withCredentials || defConfig.withCredentials;

  // 配合 bodyParser，将对象转化到 body 而非 data 上
  config.body = config.data;

  // Flatten headers
  config.headers = (0, _extends3.default)({}, config.headers.common || {}, config.headers[config.method] || {}, config.headers || {});

  ['delete', 'get', 'head', 'post', 'put', 'patch', 'common'].forEach(function (method) {
    delete config.headers[method];
  });

  config.socket = null;

  return config;
};

var getResponse = exports.getResponse = function getResponse() {
  // 这两个函数的调用在虚拟响应里没有效果
  var setHeader = function setHeader() {};
  var removeHeader = function removeHeader() {};

  return {
    statusCode: 200,
    statusMessage: 'OK',
    setHeader: setHeader,
    removeHeader: removeHeader,
    socket: null
  };
};
